oak_channel/SPEC.md

```diff
@@ -138,149 +93,80 @@ A byte encoded frame MUST consist of the following fields:
  0                   1                   2                   3
  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                            padding                            |
+|                         invocation_id                         |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                         message_length                        |
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|             length            |             flags             |
+|                          frame_number                         |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                          body_length                          |
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 |                                                               |
 +                         body bytes...                         +
 |                                                               |
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ...
 ```
 
-<!-- Diagram generated with https://www.luismg.com/protocol/, using the schema
-"padding:32,length:16,flags:16,body bytes...:64" -->
+<!-- Diagram generated with https://www.luismg.com/protocol/, using the spec
+"invocation_id:32,message_length:32,frame_number:32,body_length:32,body bytes...:64" -->
 
-- `padding`, 4 bytes
+- `invocation_id`, u32, little endian
 
-  For future use.
+  The unique ID of the invocation that the frame's forms a part of. All the
+  frames that make up a message MUST have the same invocation ID. The invocation
+  ID of a response message MUST match the value of its related request message.
+  The invocation ID MUST be incremented for each method invocation, naturally
+  wrapping when the maximum value for the data type is reached.
 
-- `length`, u16, little endian
+- `message_length`, unsigned 32-bit little-endian integer
 
-  Length of the entire frame, including this field. The length of a frame MUST
-  NOT exceed 4,000 bytes.
+  The length of the overall message. This value MUST be larger or equal to
+  `body_length`. The message length for all frames that make up the message MUST
+  be the same.
 
-- `flags`, 2 bytes
+  The size of a single message MUST NOT exceed 2GiB.
 
-  - `flags[0]`
+- `frame_number`, unsigned 32-bit little-endian integer
 
-    - 7 (MSB): unused
-    - 6: unused
-    - 5: unused
-    - 4: unused
-    - 3: unused
-    - 2: unused
-    - 1: end flag. MUST be set in the last frame of a message
-    - 0 (LSB): start flag. MUST be set in the first frame of a message
+  The position of this frame within the overall message. The first frame MUST
```

Comment:
Does the channel implementation need to assume that all of the frames in the "stream" will be maximum 4K length (in order to calculate the abount of expected frames) ?

Reply:
Yes, this is specified lower down under "Sending messages"